### ISSUE TITLE
Packages have too many files in them.

### ISSUE THREAD

**Issue by nicolasnoble** [mapbox/node-pre-gyp#356]
Starting at node-pre-gyp 0.8.0, package files are bloated. This behavior didn't occur with the previously published node-pre-gyp package, 0.7.0. The packages contain many irrelevant files for their target platform. For instance, you can download a [package built with node-pre-gyp 0.8.0](
https://storage.googleapis.com/grpc-testing-kokoro-prod/test_result_public/prod2/grpc/node/releases/macos/4/20180320-142543/grpc-precompiled-binaries/node/grpc/v1.10.1-pre1/electron-v1.8-darwin-x64-unknown.tar.gz).

This is the last package that node-pre-gyp created in our build sequence, and it contains the following listed files. If you look at all of the produced packages, it seems fairly obvious that each previous build is being aggregated into the next build's package.

```
-rwxr-xr-x  0 0      0     2649912 Mar 20 14:31 electron-v1.0-darwin-ia32-unknown/grpc_node.node
-rwxr-xr-x  0 0      0     2896120 Mar 20 14:37 electron-v1.0-darwin-x64-unknown/grpc_node.node
-rwxr-xr-x  0 0      0     2649912 Mar 20 14:31 electron-v1.1-darwin-ia32-unknown/grpc_node.node
-rwxr-xr-x  0 0      0     2896120 Mar 20 14:37 electron-v1.1-darwin-x64-unknown/grpc_node.node
-rwxr-xr-x  0 0      0     2649912 Mar 20 14:32 electron-v1.2-darwin-ia32-unknown/grpc_node.node
-rwxr-xr-x  0 0      0     2896120 Mar 20 14:38 electron-v1.2-darwin-x64-unknown/grpc_node.node
-rwxr-xr-x  0 0      0     2649936 Mar 20 14:32 electron-v1.3-darwin-ia32-unknown/grpc_node.node
-rwxr-xr-x  0 0      0     2896152 Mar 20 14:38 electron-v1.3-darwin-x64-unknown/grpc_node.node
-rwxr-xr-x  0 0      0     2649936 Mar 20 14:33 electron-v1.4-darwin-ia32-unknown/grpc_node.node
-rwxr-xr-x  0 0      0     2896152 Mar 20 14:39 electron-v1.4-darwin-x64-unknown/grpc_node.node
-rwxr-xr-x  0 0      0     2649936 Mar 20 14:33 electron-v1.5-darwin-ia32-unknown/grpc_node.node
-rwxr-xr-x  0 0      0     2896152 Mar 20 14:39 electron-v1.5-darwin-x64-unknown/grpc_node.node
-rwxr-xr-x  0 0      0     2649936 Mar 20 14:34 electron-v1.6-darwin-ia32-unknown/grpc_node.node
-rwxr-xr-x  0 0      0     2896152 Mar 20 14:40 electron-v1.6-darwin-x64-unknown/grpc_node.node
-rwxr-xr-x  0 0      0     2649936 Mar 20 14:34 electron-v1.7-darwin-ia32-unknown/grpc_node.node
-rwxr-xr-x  0 0      0     2896152 Mar 20 14:40 electron-v1.7-darwin-x64-unknown/grpc_node.node
-rwxr-xr-x  0 0      0     2649936 Mar 20 14:35 electron-v1.8-darwin-ia32-unknown/grpc_node.node
-rwxr-xr-x  0 0      0     2896152 Mar 20 14:41 electron-v1.8-darwin-x64-unknown/grpc_node.node
-rwxr-xr-x  0 0      0     1269496 Mar 20 14:29 node-v46-darwin-ia32-unknown/grpc_node.node
-rwxr-xr-x  0 0      0     1349364 Mar 20 14:35 node-v46-darwin-x64-unknown/grpc_node.node
-rwxr-xr-x  0 0      0     1269496 Mar 20 14:29 node-v47-darwin-ia32-unknown/grpc_node.node
-rwxr-xr-x  0 0      0     1349364 Mar 20 14:35 node-v47-darwin-x64-unknown/grpc_node.node
-rwxr-xr-x  0 0      0     1269496 Mar 20 14:29 node-v48-darwin-ia32-unknown/grpc_node.node
-rwxr-xr-x  0 0      0     1349364 Mar 20 14:35 node-v48-darwin-x64-unknown/grpc_node.node
-rwxr-xr-x  0 0      0     1269544 Mar 20 14:29 node-v51-darwin-ia32-unknown/grpc_node.node
-rwxr-xr-x  0 0      0     1349412 Mar 20 14:36 node-v51-darwin-x64-unknown/grpc_node.node
-rwxr-xr-x  0 0      0     1269544 Mar 20 14:30 node-v57-darwin-ia32-unknown/grpc_node.node
-rwxr-xr-x  0 0      0     1349412 Mar 20 14:36 node-v57-darwin-x64-unknown/grpc_node.node
-rwxr-xr-x  0 0      0     1270140 Mar 20 14:30 node-v59-darwin-ia32-unknown/grpc_node.node
-rwxr-xr-x  0 0      0     1350044 Mar 20 14:36 node-v59-darwin-x64-unknown/grpc_node.node
```

**Comment by springmeyer** [comment]
Hi @nicolasnoble - thanks for reporting. Maybe this has to do with something changing after #345.

Can you provide a smaller, reproducible testcase?

**Comment by nicolasnoble** [comment]
Here's a repository with a simple test case: https://github.com/nicolasnoble/node-pre-gyp-testcase

If you run `build.sh` from HEAD, you will see that the generated 64 bits package contains both 32 bits and 64 bits builds:

```
$ tar tvfz artifacts/helloworld-v1.0.0-node-*-x64.tar.gz
-rwxrwxr-x 0/0            7816 2018-03-22 00:15 node-v57-linux-ia32-glibc/helloworld.node
-rwxrwxr-x 0/0           13280 2018-03-22 00:15 node-v57-linux-x64-glibc/helloworld.node
```

But if you remove the top commit to switch back to node-pre-gyp 0.7.0, this behavior isn't present anymore.

**Comment by springmeyer** [comment]
@nicolasnoble thank you very much for the reduced case. @inspiredware is going to take a look to make sure this is not related to #345. If that does not turn anything up I'll take a closer look next week.

**Comment by inspiredware** [comment]
@springmeyer the issue is not related to the changes for N-API in #345. 
The code in `lib/package.js` that actually creates the `.tar.gz` file changed between `0.7.0` and `0.8.0` in commit bb3bdaed. It appears that the new code, at line 55 in `0.8.0`, is assuming the `extension_binary` directory will contain just a single directory with the current build product. However, the contents of the `extension_binary` directory are not cleared out between builds. Hence, the resulting the `.tar.gz` file contains all previous builds, as @nicolasnoble reports. 
The best fix is beyond my capabilities at the moment, so I need to turn it back over to you.

**Comment by springmeyer** [comment]
@inspiredware I'm hugely thankful for you narrowing down the commit. I will take a look next week.

**Comment by agathver** [comment]
@springmeyer  Tar-pack and npm-packlist seem to have different goals. Tar-pack only kept compiled node_modules, while npm-packlist appears to contain all files.

**Comment by springmeyer** [comment]
@agathver perhaps, or perhaps it is the extra `from = path.dirname(from);` that came in with bb3bdae at https://github.com/mapbox/node-pre-gyp/commit/bb3bdaed56d6540bdd9519f2dbcfdd27d99ed22e#diff-8eaf344b0d37588331d22edb952ddd05R33?

**Comment by springmeyer** [comment]
should be fixed by #371
